src/wms/src/main/java/org/geoserver/wms/map/RenderedImageMapOutputFormat.java

```diff
@@ -505,57 +509,86 @@ public RenderedImageMap produceMap(final WMSMapContent mapContent, final boolean
         
         onBeforeRender(renderer);
 
-        // setup the timeout enforcer (the enforcer is neutral when the timeout is 0)
+        //Timeout on the smallest nonzero value of the WMS timeout and the timeout format option
+        //If both are zero then there is no timeout
         int maxRenderingTime = wms.getMaxRenderingTime() * 1000;
-        RenderingTimeoutEnforcer timeout = new RenderingTimeoutEnforcer(maxRenderingTime, renderer,
-                graphic);
-        timeout.start();
-        try {
-            // finally render the image;
-            renderer.paint(graphic, paintArea, mapContent.getRenderingArea(),
-                    mapContent.getRenderingTransform());
-
-            // apply watermarking
-            if (layout != null) {
-                try {
-                    layout.paint(graphic, paintArea, mapContent);
-                } catch (Exception e) {
-                    throw new ServiceException("Problem occurred while trying to watermark data", e);
-                }
+        int localMaxRenderingTime = 0;
+        
+        Object timeoutOption = request.getFormatOptions().get("timeout");
+        if (timeoutOption != null) {
+            try {
+                localMaxRenderingTime = Integer.parseInt(timeoutOption.toString());
+            } catch (NumberFormatException e) {
+                LOGGER.log(Level.WARNING,"Could not parse format_option \"timeout\": "+timeoutOption, e);
             }
-        } finally {
-            timeout.stop();
-            graphic.dispose();
         }
-
-        // check if the request did timeout
-        if (timeout.isTimedOut()) {
-            throw new ServiceException(
-                    "This requested used more time than allowed and has been forcefully stopped. "
-                            + "Max rendering time is " + (maxRenderingTime / 1000.0) + "s");
+        if (maxRenderingTime == 0) {
+            maxRenderingTime = localMaxRenderingTime;
+        } else if (localMaxRenderingTime != 0) {
+            maxRenderingTime = Math.min(maxRenderingTime, localMaxRenderingTime);
         }
-
-        // check if a non ignorable error occurred
-        if (nonIgnorableExceptionListener.exceptionOccurred()) {
-            Exception renderError = nonIgnorableExceptionListener.getException();
-            throw new ServiceException("Rendering process failed", renderError, "internalError");
+        
+        //Run rendering in a separate thread so that we can exit promptly if the timeout is reached,
+        //rather than waiting for the renderer to finish the current task.
+        MapPainter painter = new MapPainter(renderer, paintArea, mapContent, graphic, layout);
```

Comment:
Good to get some input from someone who knows the internals of the renderer better than I do. I'll move back to the previous method of termiating the rendering process.

From what I have observed, it looks like renderer will not stop rendering in the middle of drawing one feature. One example is the natural earth oceans layer (which is one big feature) - if we start rendering this, the renderer will try to render the whole thing (or at least the first symbolizer) regardless of timeout. Is there anything that can be done about this? A more reasonable case is some more ordinary layer that just happens to have some very large features in it - such a layer will not respect any timeout very well.